jsparagus/parse_table.py

```diff
@@ -1314,6 +1322,138 @@ def remove_all_unreachable_state(self, verbose: bool, progress: bool) -> None:
         self.states = [s for s in self.states if s is not None]
         self.rewrite_reordered_state_indexes()
 
+    def remove_reduce(self, verbose: bool, progress: bool) -> None:
+        # Remove Reduce actions and replace them by the programmatic
+        # equivalent.
+        #
+        # This transformation preserves the stack manipulations of the parse
+        # table. It only changes it from being implicitly executed by the shift
+        # function of the LR parser, by being executed as an action.
+        #
+        # This transformation converts the hard-to-predict load of the shift
+        # table into a branch prediction which is potentially easier to
+        # predict.
+        #
+        # A side-effect of this transformation is that it removes the need for
+        # replaying non-terminals, thus the backends could safely ignore the
+        # ability of the shift function from handling non-terminals.
+        if verbose or progress:
+            print("Remove Reduce actions.")
+
+        maybe_unreachable_set: OrderedSet[StateId] = OrderedSet()
+
+        def transform() -> typing.Iterator[None]:
+            for s in self.states:
+                term, _ = next(iter(s.epsilon), (None, None))
+                if self.term_is_shifted(term):
+                    continue
+                assert len(s.epsilon) == 1
+                yield  # progress bar.
+                reduce_state = s
+                if verbose:
+                    print("Inlining shift-operation for state {}".format(str(reduce_state)))
+
+                # The reduced_aps should contain all reduced path of the single
+                # Reduce action which is present on this state. However, as
+                # state of the graph are shared, some reduced paths might follow
+                # the same path and reach the same state.
+                #
+                # This code collect for each replayed path, the tops of the
+                # stack on top of which these states are replayed.
+                aps = APS.start(s.index)
+                states_by_replay_term = collections.defaultdict(list)
+                # print("Start:\n{}".format(aps.string(name="\titer_aps")))
+                # print(s.stable_str(self.states))
+                for reduced_aps in aps.shift_next(self):
+                    # As long as we have elements to replay, we should only
+                    # have a single path for each reduced path. If the next
+                    # state contains an action, then we stop here.
```

Comment:
To clarify, "stop here" means that at run time we'll leave some tokens in the actual runtime replay buffer (we sort of "spill" them, we're unable to implement the remaining replay actions in straight line code), but the behavior is unchanged. Right?